logging: add openamp rpmsg backend

nslowell opened the pull request:
Add backend that sends logs over an RPMsg endpoint

Remoteproc system with an established openamp rpmsg communication connection can acquire logging information over a dedicated endpoint.

sylvioalves commented on subsys/logging/backends/Kconfig.rpmsg:
```diff
@@ -0,0 +1,52 @@
+	default 256
+	help
+	  In deferred logging mode, sets the maximum number of bytes which can be buffered in
+	  RAM before log_output_flush is automatically called on the RPMSG backend.  The buffer
```
Extra space typo at `backend.  The buffer`

sylvioalves commented on subsys/logging/backends/log_backend_rpmsg.c:
```diff
@@ -0,0 +1,201 @@
+				void *priv)
+{
+	/* No incoming messages expected - just return success */
+
```
Would you add `ARG_UNUSED()` for all arguments in here?

sylvioalves commented on subsys/logging/backends/log_backend_rpmsg.c:
```diff
@@ -0,0 +1,201 @@
+	struct log_backend_rpmsg_data *backend_data = backend_ctx->data;
+	int ret = 0;
+
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_RPMSG_BLOCKING_MODE)) {
```
Please check whether `CONFIG_LOG_MODE_IMMEDIATE` can even be used when `CONFIG_LOG_BACKEND_RPMSG_BLOCKING_MODE` is enabled. In that cause, I think some build warning (or Kconfig handling) may be necessary.

uLipe reviewed (commented):
Nice work! I just have a suggestion, not a blocker :)

sylvioalves commented on subsys/logging/backends/log_backend_rpmsg.c:
```diff
@@ -0,0 +1,201 @@
+	log_output_func = log_format_func_t_get(data->log_format_current);
+
+	/* Format and output the message */
+	log_output_func(ctx->output, &msg->log, flags);
```
Check for `log_output_func != NULL` if needed.

arnopo commented on samples/subsys/ipc/openamp_rsc_table/prj.conf:
```diff
@@ -10,3 +10,7 @@ CONFIG_OPENAMP_MASTER=n
 
 CONFIG_SHELL=y
 CONFIG_SHELL_BACKEND_RPMSG=y
+CONFIG_SHELL_MINIMAL=y # Enabled to include Shell and Log for stm32mp157c_dk2 board
```
Any reason to specify the board here? Is it the board use for your test?

arnopo replied:
1) This looks like a nice feature, although I am not very familiar with its usage. How does this coexist with the Shell already implemented? 

2) Regarding the sample itself
The sample seems already quite complex as it already integrates:
- Two TTY rpmsg devices (one for echo, one for the shell)
- one responder for the Linux client sample

Unless I missed something, this will create a third TTY rpmsg device on the Linux side. I wonder if it would be better to create a new sample for the feature instead of making this one overly complex. Notice that is not a request but a question for maintainers.

Additionally, I would expect to find information in the README file explaining how to run the sample.

Commit:
logging: add openamp rpmsg backend

Add backend that sends logs over an RPMsg endpoint

Remoteproc system with an established openamp rpmsg
communication connection can acquire logging information
over a dedicated endpoint.

Signed-off-by: Nicholas Lowell <Nicholas.Lowell@lexmark.com>

Commit:
samples: logging: rpmsg_backend example application

sample to show how to use the rpmsg logging backend
after establishing an OpenAMP connection

intended for platforms that embed Linux OS on main core
and Zephyr on a co-processor

Signed-off-by: Nicholas Lowell <Nicholas.Lowell@lexmark.com>

sylvioalves commented on include/zephyr/logging/log_backend_rpmsg.h:
```diff
@@ -0,0 +1,33 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LOG_BACKEND_RPMSG_H__
```
Would you keep a single underscore at the end to meet common code style?

nslowell replied:
none of the checks tagged this...will fix.

iuliana-prodan commented on subsys/logging/backends/log_backend_rpmsg.c:
```diff
@@ -0,0 +1,223 @@
+		return ret;
+	}
+
+	return (int)length;
```
Returning the full length on error might hide issues.
Please, consider logging the error or implementing a retry mechanism.

arnopo commented on include/zephyr/logging/log_backend_rpmsg.h:
```diff
@@ -0,0 +1,33 @@
+#define ZEPHYR_LOG_BACKEND_RPMSG_H_
+
+#include <zephyr/logging/log_backend.h>
+#include <openamp/rpmsg.h>
```
nitpicking, prefer alphabetical ordering

arnopo commented on subsys/logging/backends/log_backend_rpmsg.c:
```diff
@@ -0,0 +1,223 @@
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/sys/__assert.h>
+
+#include <openamp/rpmsg.h>
```
same includes to be reordered

arnopo commented on subsys/logging/backends/log_backend_rpmsg.c:
```diff
@@ -0,0 +1,223 @@
+	/* Return number of bytes sent or full length if error to avoid potential forever loop */
+	if (ret >= 0) {
+		return ret;
+	}
```
Here we have a silence drop, right?
Could we call log_backend_std_dropped in such case and return 0?
I assume we cannot use LOG_ERR here to indicate the issue...

iuliana-prodan commented on samples/subsys/logging/rpmsg_backend/boards/imx8mp_evk_mimx8ml8_adsp.overlay:
```diff
@@ -0,0 +1,24 @@
+ */
+
+/ {
+	chosen {
```
Please keep only imx8mp_evk/mimx8ml8/adsp — I've verified this and it's fine.
If you haven’t tested the other boards (I didn’t have time to check each one myself), please remove their overlays.
Also in `sample.yaml` you have only these platforms allowed:    
      - stm32mp157c_dk2
      - imx8mp_evk/mimx8ml8/adsp
So, keep only what was tested.

JarmouniA commented on samples/subsys/logging/rpmsg_backend/src/main.c:
```diff
@@ -0,0 +1,220 @@
+	vdev = rproc_virtio_create_vdev(VIRTIO_DEV_DEVICE, VDEV_ID, rsc_table_to_vdev(rsc_table),
+					rsc_io, NULL, mailbox_notify, NULL);
+
+	if (!vdev) {
```
```suggestion
	if (vdev == NULL) {
```

JarmouniA commented on samples/subsys/logging/rpmsg_backend/src/main.c:
```diff
@@ -0,0 +1,220 @@
+	}
+
+	rpdev = platform_create_rpmsg_vdev(0, VIRTIO_DEV_DEVICE, NULL, new_service_cb);
+	if (!rpdev) {
```
```suggestion
	if (rpdev == NULL) {
```

nslowell commented on samples/subsys/logging/rpmsg_backend/src/main.c:
```diff
@@ -0,0 +1,220 @@
+		printk("Sending message...\n");
+		/* Send to all log backends */
+		LOG_INF("OpenAMP[remote] Hello Linux!");
+	}
```
The default output from the log has a timestamp appended already, is that not dynamic enough?

```
   [00:00:09.001,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!
   Sending message...
   [00:00:11.002,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!
   Sending message...
   [00:00:13.002,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!
   Sending message...
   [00:00:15.003,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!
```

iuliana-prodan commented:
@nslowell this looks mostly good!

However, when I tested on i.MX8M Plus with DSP, the DSP console was flooded with failed messages until Linux (driver side) sent a ping (or any message).
```
*** Booting Zephyr OS build v4.3.0-rc1-77-g612ce992559b ***
Message received
RPMSG log backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
...
RPMSG log backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
RPMSG log baSending message...
ckend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
...
RPMSG log backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2003
RPMSG logMessage received  ===========================> after Linux-Zephyr handshake is done
 backend: message send failed, err -2003
RPMSG log backend: message send failed, err -2002
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
```
To handle this RPMSG handshake issue more elegantly, I see two options:
1. modify the Zephyr code to wait for a specific message from Linux before starting to send logs
2. update the Log Backend to detect when the connection is ready

I recommend _option 2_ because it’s the most practical. 
Any message from Linux (even an empty one) means the connection is ready, so the callback can set a flag to enable normal logging.

iuliana-prodan commented:
> @iuliana-prodan Thanks for continuing to check on this. I think I've got it buttoned up finally.

The current change is a good improvement that makes the backend self-activating, but I still get a lot of "Sending message..." in Zephyr console, before any message from Linux.

Here's my Linux console:
```
root@imx8mpevk:~# echo -n zephyr.elf > /sys/class/remoteproc/remoteproc0/firmware
root@imx8mpevk:~# echo start > /sys/class/remoteproc/remoteproc0/state
root@imx8mpevk:~# cat /dev/ttyRPMSG0 &
[1] 1609
root@imx8mpevk:~# 
root@imx8mpevk:~# 
root@imx8mpevk:~# 
root@imx8mpevk:~# 
root@imx8mpevk:~# echo '' > /dev/ttyRPMSG0
root@imx8mpevk:~# [00:00:35.034,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:37.036,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:39.037,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:41.039,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:43.040,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:45.042,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:47.043,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:49.045,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:51.046,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:53.048,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:55.049,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:57.051,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:00:59.052,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:01:01.054,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:01:03.055,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:01:05.057,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:01:07.058,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:01:09.060,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:01:11.061,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:01:13.063,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:01:15.064,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:01:17.066,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

[00:01:19.067,000] <inf> rpmsg_backend: OpenAMP[remote] Hello Linux!

^C
root@imx8mpevk:~# echo stop > /sys/class/remoteproc/remoteproc0/state
root@imx8mpevk:~# 
```

Here's my Zephyr console:
```
*** Booting Zephyr OS build v4.3.0-438-ge8ce054c2b20 ***
Message received ======== > Why is this here?
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Message received =========> This is when I sent the echo ''  from Linux
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
Sending message...
```

Why do we have the first `Message received` from Zephyr?
At first I thought that Linux sends automatic handshake, next the `log_backend_rpmsg_cb()` enables the backend, but logs aren't flowing - something is preventing the logs from being sent via RPMSG.
Without sending `echo '' > /dev/ttyRPMSG0` I see no `<inf> rpmsg_backend: OpenAMP[remote] Hello Linux!` in Linux. 
So, what am I missing?

IMO, Zephyr code should wait for a specific message from Linux before starting to send logs.

iuliana-prodan commented:
1. > Hey @iuliana-prodan

> 
> First, the obvious: "Message received" and "Sending message" are printks coming from the sample application. Actual implementations could choose differently when/how to send messages to the log.
> 
> Next, the answers to your questions: Here's the sequence of events:
> 
> 1. Linux remoteproc driver loads firmware
> 2. Linux remoteproc driver allocates vring virtqueues based on the resource table
> 3. Linux virtio_rpmsg_bus driver probes, finds the virtqueues, allocates buffers and kicks the remote processor to tell it the rpmsg bus is open for communication.   It's just a kick, no message is sent over the bus.   So your mailbox/IPM device is getting a receive interrupt, which calls the callback that prints "Message Received".
> 4. The remote processor (i.e. Zephyr sample code) nows sends its first message over the rpmsg--invoked in `log_backend_rpmsg_init_transport()`, which calls `rpmsg_create_ept()`--which is an announcement of the creation of the "rpmsg-tty" endpoint for the Linux nameservice to detect
> 5. Linux rpmsg_ns gets the message and creates /dev/ttyRPMSG0 and then rpmsg_tty driver probes.  This is the step in which the addresses change from ANY (i.e. -1) to a designated integer on the host side, but it is not communicated to the remote processor.
> 6. On the first message over the new endpoint from host to remote processor, the address is updated.  This is a key step.  Taken from open-amp/lib/rpmsg/rpmsg_virtio.c:
> 
> ```
> rpmsg_virtio_rx_callback()
> {
> ...
> 	if (ept->dest_addr == RPMSG_ADDR_ANY) {
> 		/*
> 			* First message received from the remote side,
> 			* update channel destination address
> 			*/
> 		ept->dest_addr = rp_hdr->src;
> 	}
> ...
> }
> ```
> 
> and _that_ update of the `dest_addr` for the endpoint is what finally allows `rpmsg_send()` to succeed when sending messages.
> 
> To summarize, the first "Message received" is a notify from the host when the rpmsg _bus_ is available. `log_rpmsg_backend_cb()` does not get called at that time. Zephyr then creates and announces the rpmsg endpoint channel. Only after receiving an rpmsg message from the host (Linux) over that new endpoint channel is the endpoint destination address properly updated on the Zephyr side. `log_rpmsg_backend_cb()` is called when that message is received and enables the backend. `rpmsg_send()` is now able to send log messages over the bus, through the endpoint channel.
> 
> I hope that helps. It is definitely a complex setup. I'm not sure I can change anything, this just seems to be the way the protocol is setup. I can adjust the sample if something bugs you, but it's just a sample and real implementations would probably be a little different anyways.

Thank you for the detailed explanation. However, I still believe we can improve this by modifying the Zephyr code to wait for a specific message from Linux before starting to send logs.
Here’s what I have in mind:
1. in main(), after setting up rpmsg but before initializing the log backend:
 - create a temporary rpmsg endpoint to wait for the Linux "ready" signal.
2. wait for the Linux ready signal:
- implement a mechanism to listen for the signal on the temporary endpoint.
- include a timeout to avoid blocking indefinitely.
3. once the Linux ready signal is received (or the timeout occurs):
- proceed to initialize the log backend.

That said, we can consider these improvements after merging this PR.

iuliana-prodan commented on samples/subsys/logging/rpmsg_backend/src/main.c:
@arnopo @nslowell maybe we should not use IPM at all since this is deprecated. I suggest to use mbox API.
Also, see this open discussion https://github.com/zephyrproject-rtos/zephyr/discussions/99529

nslowell reviewed (commented):
See my other comment replies.  Mostly, I'm wondering if this sample is even needed since I've now found two other samples that have almost verbatim the same code.
